### DISCUSSION TITLE
how to count based on the last relation item

### DISCUSSION THREAD

**Discussion by vitordino** [prisma/prisma#4298] [Q&A] [answered]
hey there, i stumbled on a tricky thing while doing a POC..

> it’s about _users_ doing _challenges_ to increase planted _tree_ 🌲 counts

### the schema context:

1. `User`s have `Challenge`s (_one-to-many relation_)
2. `Challenge`s have `ChallengeChange`s (_one-to-many relation_)
3. `ChallengeChange` have keys:
  a. `date` (self-explanatory)
  b. `type` (enum for: `enrolled`, `canceled` or `completed`)

### what i want

the business logic says that it should **count a tree every challenge done**.

we should count a challenge as **done** either if:
- the user marks a challenge as `completed` via a `ChallengeChange` linked to it, and that’s the last change of the challenge.

- the user started the challenge (via a `enrolled` type `ChallengeChange`) more than 7 days ago, and that’s the last change of the challenge (as they can go back and tick it `cancelled` anytime)

### what i tried:

```ts
const countTrees = (challengeKey: string) => {
  const now = new Date().getTime()
  const timeLimit = new Date(now - TIME_OFFSET)

  return prisma.challenge.count({
    where: {
      AND: [
        { name: challengeKey },
        {
          OR: [
            { changes: { some: { type: 'COMPLETE' } } },
            { changes: { some: { type: 'ENROLL', date: { lte: timeLimit } } } },
          ],
        },
      ],
    },
    distinct: 'id',
  })
}
```

### problem

obviously i’m counting all the `changes` each `challenge` took (that’s not only wrong on the _business logic_ but also not great performance-wise).

it would be great if i could select only the last change of each challenge (either automatically, or using `orderBy`)

**Comment by vitordino** [comment]


**Reply by ryands17** [reply, marked as answer]
Hey @vitordino 👋 
Thanks for creating this discussion, an interesting use-case indeed. Unfortunately this won't be possible directly with Prisma's inbuilt API for models, so you would need to stick to using [this](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access) raw query via Prisma.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
